src/session_server_ssh.c

```diff
@@ -397,110 +415,279 @@ nc_pam_conv_clb(int n_messages, const struct pam_message **msg, struct pam_respo
+{
+    int ret = 0, n_answers;
+    const char *name = "Keyboard-Interactive Authentication";
+    const char *instruction = "Please enter your authentication token";
+    char *prompt = NULL, *local_pw = NULL, *received_pw = NULL;
+    char echo[] = {0};
+
+    /* try to get the client's locally stored pw hash */
+    local_pw = nc_server_ssh_get_pwd_hash(client->username);
+    if (!local_pw) {
+        ERR(session, "Unable to get %s's credentials.", client->username);
+        ret = 1;
+        goto cleanup;
+    }
+
+    ret = asprintf(&prompt, "%s's password:", client->username);
+    NC_CHECK_ERRMEM_GOTO(ret == -1, prompt = NULL; ret = 1, cleanup);
+
+    /* send the password prompt to the client */
+    ret = ssh_message_auth_interactive_request(msg, name, instruction, 1, (const char **) &prompt, echo);
+    if (ret) {
+        ERR(session, "Failed to send an authentication request to client \"%s\".", client->username);
+        goto cleanup;
+    }
+
+    /* get the reply */
+    n_answers = nc_server_ssh_kbdint_get_nanswers(session, session->ti.libssh.session, auth_timeout);
+    if (n_answers < 0) {
+        /* timeout or dc */
+        ret = 1;
+        goto cleanup;
+    } else if (n_answers != 1) {
+        /* only expecting a single answer */
+        ERR(session, "Unexpected amount of answers in system auth. Expected 1, got \"%d\".", n_answers);
+        ret = 1;
+        goto cleanup;
+    }
+    received_pw = strdup(ssh_userauth_kbdint_getanswer(session->ti.libssh.session, 0));
+    NC_CHECK_ERRMEM_GOTO(!received_pw, ret = 1, cleanup);
+
+    /* cmp the pw hashes */
+    ret = auth_password_compare_pwd(local_pw, received_pw);
+
+cleanup:
+    free(local_pw);
+    free(received_pw);
+    free(prompt);
+    return ret;
+}
+
 #endif
 
 static int
-nc_sshcb_auth_kbdint(struct nc_session *session, struct nc_server_ssh_opts *opts, ssh_message msg)
+nc_sshcb_auth_kbdint(struct nc_session *session, struct nc_auth_client *client, uint16_t auth_timeout, ssh_message msg)
 {
     int auth_ret = 1;
 
     if (server_opts.interactive_auth_clb) {
         auth_ret = server_opts.interactive_auth_clb(session, session->ti.libssh.session, msg, server_opts.interactive_auth_data);
     } else {
 #ifdef HAVE_LIBPAM
-        if (nc_pam_auth(session, opts, msg) == PAM_SUCCESS) {
+        /* authenticate using PAM */
+        if (!nc_pam_auth(session, client, auth_timeout, msg)) {
+            auth_ret = 0;
+        }
+#elif defined (HAVE_SHADOW)
+        /* authenticate using locally configured users */
+        if (!nc_server_ssh_system_auth(session, client, auth_timeout, msg)) {
             auth_ret = 0;
         }
 #else
-        ERR(session, "PAM-based SSH authentication is not supported.");
+        (void) auth_timeout;
+        ERR(NULL, "No Keyboard-interactive mechanisms to authenticate the client \"%s\" found.", client->username);
```

Comment:
I would expect something shorter such as `Keyboard-interactive authentication method not supported.`.